### DISCUSSION TITLE
Lesson 11 initial test error question

### DISCUSSION THREAD

**Discussion by RichardPalladino** [Cyfrin/foundry-full-course-cu#434] [Q&A] [answered]
This isn't so much a question about Lesson 11 as it is about Solidity / Foundry peculiarities.

The first test, `testNameIsCorrect`, should be simple enough but I had it fail when I used assertEq instead of just assert.

My initial code:
```
    function testNameIsCorrect() public view {
        string memory expectedName = "Doggie";
        string memory actualName = basicNft.name();
        assertEq(keccak256(abi.encodePacked(actualName)), keccak256(abi.encodePacked(expectedName)));
    }
```
failed with the error:
```
Error (8961): Function cannot be declared as view because this expression (potentially) modifies the state.
  --> test/unit/BasicNft.t.sol:21:9:
   |
21 |         assertEq(keccak256(abi.encodePacked(actualName)), keccak256(abi.encodePacked(expectedName)));
```

When I used assert like below, it works:
```
    function testNameIsCorrect() public view {
        string memory expectedName = "Doggie";
        string memory actualName = basicNft.name();
        assert(keccak256(abi.encodePacked(actualName)) == keccak256(abi.encodePacked(expectedName)));
    }
```

To me, they both should work.  What error or issue am I missing?

**Comment by RichardPalladino** [comment]
Follow up:  it works when I remove "view" from the function.

I just feel like it should work either way.  What makes the compiler think `assertEq` could change modify state while `assert` is perfectly okay?

**Comment by shikhar229169** [comment, marked as answer]
When you will command click assertEq you will find out that it actually emits event in case if the left and right values were not equal. So if a function emits an emit, then it actually modifies the state.

**Reply by RichardPalladino** [reply]
Function emits modify state?  Wow.  That's new to me.  Thanks for educating me.  I always thought an emit was like logging, something that actually did not alter state in any way.

**Comment by RichardPalladino** [comment]
https://ethereum.stackexchange.com/questions/132618/why-is-emitting-an-event-considered-as-state-changing

**Reply by RichardPalladino** [reply]
They don't change the contract state, but change the blockchain state and therefore are considered state-changing.  Super interesting.
